src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs

```diff
@@ -512,14 +512,17 @@ public virtual Expression BindInNode(InNode inNode)
             Expression singleValue = Bind(inNode.Left);
             Expression collection = Bind(inNode.Right);
 
+            Type collectionItemType = collection.Type.GetElementType() ?? (typeof(IEnumerable).IsAssignableFrom(collection.Type) ? collection.Type.GetGenericArguments()[0] : null);
```

Comment:
>typeof(IEnumerable).IsAssignableFrom(collection.Type) [](start = 75, length = 53)

So, the assertion is that either the collection expression returns a type with a non-null GetElementType() or it is an IEnumerable<T>, right? is that the contract of Bind() for any CollectionNode?

Note that, if collection is a non-generic IEnumerable you will get an out of range exception when try to index into the first member of GetGenericArguments() before you get to your Contract.Assert --so that assert is not doing much and neither, really, is checking that typeof(IEnumerable).IsAssignableFrom(collection.Type).

Consider breaking up, something like:
            Type collectionItemType = collection.Type.GetElementType();
            if (collectionItemType == null)
            {
                Type[] genericArgs = collection.Type.GetGenericArguments();
                Contract.Assert(genericArgs.Length == 1);
                collectionItemType = collection.Type.GetGenericArguments()[0];
            }

If we aren't guaranteed that Bind() will return a non-null GetElementType() or IEnumerable<T> we should change the Contract.Assert into an explicit error case.

Reply:
@mikepizzo I implemented the changes. Presently, the model builder does not support non-generic collections like `ArrayList` or generic collections with generic arguments > 1 like `IDictionary<,>` so a `Contract.Assert()` would suffice for now?